### PR TITLE
WT-6479 Don't insert updates after the first globally visible update to the history store

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

File: src/history/hs.c [hunk 1]
```diff
@@ -799,27 +799,6 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
...
-                __wt_verbose(session, WT_VERB_TIMESTAMP,
-                  "fixing out-of-order updates during insertion; start_ts=%s, durable_start_ts=%s, "
-                  "min_insert_ts=%s",
-                  __wt_timestamp_to_string(upd->start_ts, ts_string[0]),
-                  __wt_timestamp_to_string(upd->durable_ts, ts_string[1]),
-                  __wt_timestamp_to_string(min_insert_ts, ts_string[2]));
-                upd->start_ts = upd->durable_ts = min_insert_ts;
-                WT_STAT_CONN_INCR(session, cache_hs_order_fixup_insert);
-            } else
-                min_insert_ts = upd->start_ts;
```
Comment (agorrod) [inline src/history/hs.c:822]: Would a different approach to this change be to only save `min_insert_ts` if `upd->start_ts` is not `WT_TS_NONE`?
Is that what this change is about? We don't want to reset timestamps in the past in the special case of mixed-mode timestamps? I think it's worrying that something is relying on that semantic. Does it mean that a transaction-ID based check is causing the globally visible update to be skipped, then relying on a timestamp check to skip seeing the first update that was older than it?

Comment (quchenhao) [reply src/history/hs.c:822]: I'm discussing with Vamsi about a better solution for this problem. This PR actually doesn't handle all the edge cases and I think there are better ways.

Comment (quchenhao) [reply src/history/hs.c:822]: The problem is previously we skipping inserting updates after a mixed mode operation but WT-6388 is changing the timestamps after the mixed mode operation to zero so we can't detect mixed mode operation anymore, thus trigger the assert.

Comment (quchenhao) [reply src/history/hs.c:822]: The best solution I have in mind is to let obsolete check tell us the first globally visible update and skip inserting everything after that.

File: src/history/hs.c [hunk 2]
```diff
@@ -698,7 +698,8 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
     WT_MODIFY entries[MAX_REVERSE_MODIFY_NUM];
     WT_MODIFY_VECTOR modifies;
     WT_SAVE_UPD *list;
-    WT_UPDATE *first_non_ts_upd, *non_aborted_upd, *oldest_upd, *prev_upd, *upd;
+    WT_UPDATE *first_globally_visible_upd, *first_non_ts_upd, *non_aborted_upd, *oldest_upd,
+      *prev_upd, *upd;
```
Comment (agorrod) [inline src/history/hs.c:702]: We don't wrap parameter lists like this. Add a new type declaration on a new line. Having enough local variables to need wrapping is an indication that the function is overloaded.

File: src/history/hs.c [hunk 3]
```diff
@@ -799,29 +800,14 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
                 continue;
 
             non_aborted_upd = upd;
-
-            /* If we've seen a smaller timestamp before, use that instead. */
-            if (min_insert_ts < upd->start_ts) {
```
Comment (agorrod) [inline src/history/hs.c:804]: I'd rather this code stayed here, and handled WT_TS_NONE specially, than it is moved below.

I'm concered with the code movement that some (pathological) combination of mixed mode and out of order timestamps will mean that we don't fixup out-of-order timestamps.

File: src/history/hs.c [hunk 4]
```diff
@@ -838,19 +824,50 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
...
                 if (F_ISSET(upd, WT_UPDATE_HS))
                     ts_updates_in_hs = true;
             }
 
+            /*
+             * No need to continue if we see the first self contained value after the first globally
+             * visible value.
+             */
+            if (first_globally_visible_upd != NULL && WT_UPDATE_DATA_VALUE(upd))
+                break;
```
Comment (agorrod) [inline src/history/hs.c:869]: This short circuit means we can't rely on updates older than this being marked with `WT_UPDATE_OBSOLETE` - is that OK?

Should we be careful at all about updates that come into this function already marked as obsolete?

File: src/history/hs.c [hunk 5]
```diff
@@ -975,6 +977,14 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
                 continue;
             }
 
+            /* Clear history store content if we skip inserting the updates without timestamp. */
+            if (clear_hs && upd->start_ts != WT_TS_NONE) {
+                /* We can only delete history store entries that have timestamps. */
```
Comment (agorrod) [inline src/history/hs.c:982]: Why?

File: src/include/btmem.h [hunk 6]
```diff
@@ -1075,7 +1075,7 @@ struct __wt_update {
 
 /* AUTOMATIC FLAG VALUE GENERATION START */
 #define WT_UPDATE_HS 0x01u                       /* Update has been written to history store. */
-#define WT_UPDATE_MASKED_BY_NON_TS_UPDATE 0x02u  /* Update masked by updates without timestamp. */
+#define WT_UPDATE_OBSOLETE 0x02u                 /* Update that is obsolete. */
```
Comment (agorrod) [inline src/include/btmem.h:1078]: The indenting here doesn't look right.

File: src/history/hs.c (hunk 2 above)
Comment (quchenhao) [reply src/history/hs.c:702]: Yes, I agree it is overloaded. I'll see if I can move some logic to another function.

File: src/history/hs.c (hunk 4 above)
Comment (quchenhao) [reply src/history/hs.c:869]: That is OK as they are not pushed to the stack. We only care about updates pushed to the stack after this loop.

File: src/history/hs.c (hunk 5 above)
Comment (quchenhao) [reply src/history/hs.c:982]: In a case that U2@20 -> U1@0 and U2 is globally visible, we skip inserting U1 which is a mixed mode operation. We need to clear the history store before U2 is inserted.

File: src/history/hs.c (hunk 3 above)
Comment (quchenhao) [reply src/history/hs.c:804]: Understood.

Comment (quchenhao) [reply src/history/hs.c:804]: I think the logic should be we fix all the out of order timestamp before any mixed mode operation and ignore all the updates with timestamps after we see the first mixed mode operation. So I moved the logic there.

File: src/history/hs.c [hunk 7]
```diff
@@ -838,19 +824,50 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
...
+                        WT_STAT_CONN_INCR(session, cache_hs_order_lose_durable_timestamp);
+                    }
+                    __wt_verbose(session, WT_VERB_TIMESTAMP,
+                      "fixing out-of-order updates during insertion; start_ts=%s, "
+                      "durable_start_ts=%s, "
+                      "min_insert_ts=%s",
+                      __wt_timestamp_to_string(upd->start_ts, ts_string[0]),
+                      __wt_timestamp_to_string(upd->durable_ts, ts_string[1]),
+                      __wt_timestamp_to_string(min_insert_ts, ts_string[2]));
+                    upd->start_ts = upd->durable_ts = min_insert_ts;
```
Comment (bvpvamsikrishna) [inline src/history/hs.c:850]: if we are changing durable_ts, don't we change the version stability i.e. versions newer than stable timestamp might become stable, if we move the durable_ts from greater than stable timestamp to lesser than the stable timestamp

Comment (quchenhao) [reply src/history/hs.c:850]: If we need to change the timestamp, that means there is a version with the smaller timestamp newer than us. That update should already be stable.

File: src/include/btmem.h (hunk 6 above)
Comment (quchenhao) [reply src/include/btmem.h:1078]: I don't see any problem here. Can you elaborate?

File: src/history/hs.c (hunk 3 above)
Comment (quchenhao) [reply src/history/hs.c:804]: Done

File: src/history/hs.c (hunk 2 above)
Comment (quchenhao) [reply src/history/hs.c]: Created WT-6489 to refactor it.

File: src/history/hs.c (hunk 7 above)
Comment (quchenhao) [reply src/history/hs.c]: It may be a problem but I won't change it in this PR.

File: src/include/btmem.h (hunk 6 above)
Comment (agorrod) [reply src/include/btmem.h:1078]: This is probably my past coming back to haunt me. clang should get it right.

**Pull request merged**